torch/_ops.py

```diff
@@ -25,6 +24,79 @@ def dl_open_guard():
     if _SET_GLOBAL_FLAGS:
         sys.setdlopenflags(old_flags)
 
+class OpOverload:
+    def __init__(self, op, schema):
+        self.op = op
+        self.schema = schema
+
+    def __deepcopy__(self, memo=None):
+        return self
+
+    def __call__(self, *args, **kwargs):
+        return self.op(*args, **kwargs or {})
+
+    def __getattr__(self, key):
+        return getattr(self.op, key)
+
+    # `my_namespace::my_op`
+    def qualified_name(self):
+        return self.schema.name
+
+    def overload_name(self):
+        return self.schema.overload_name
+
+    # vector of Arguments -- each Argument's name, and type can currently be accessed
+    # it might be useful to make the mutability of the arg accessible (using the alias_info stored in the C++ struct)
+    def returns(self):
+        return self.schema.returns
+
+    def arguments(self):
+        return self.schema.arguments
+
+class OpOverloadPacket():
+    def __init__(self, qualified_op_name, op_name, op):
+        self.qualified_op_name = qualified_op_name
+        self.op_name = op_name
+        self.op = op
+
+    def __deepcopy__(self, memo=None):
+        return self
+
+    # def __repr__(self):
+    #     return f"OpOverloadPacket({self.op_name}, {id(self)})"
+
+    def __getattr__(self, key):
+        # It is not a valid op_name when __file__ is passed in
+        if key == '__file__':
+            return 'torch.ops'
+        try:
+            use_key = "" if key == 'default' else key
+            op_ = torch._C._get_operation_overload(self.qualified_op_name, use_key)
+            schema = torch._C._get_schema(self.qualified_op_name, use_key)
+            overload = OpOverload(op_, schema)
+            # cache the overload object
+            setattr(self, key, overload)
+            return overload
+        except RuntimeError:
+            out = getattr(self.op, key)
+            return out
+
+    def __call__(self, *args, **kwargs):
+        # to ensure torch.ops.foo.bar() is still callable from JIT
+        # "" key handling from above
+        # save a fn_ptr and just call that here. in its current version, it will be pretty slow
+        if kwargs is None:
+            kwargs = {}
+        return self.op(*args, **kwargs)
+
+# resolution of torch.fn is different from torch.ops. ... fn
+# First one is parsing the python objects and then matching the schema -> calls into the unboxed version of the method
+
+# second one is done by JIT. Creates a stack of all the overloads and then tries to match the correct one at runtime
+# and always calls into boxed version of the method
+
+# autograd codegen mostly creates variabletype, tracertype, inplace or view type and python bindings
```

Comment:
Are these comments meant to remain?

Reply:
yeah I think we should keep these comments for future reference

Reply:
Ok, I think they should be cleaned up then, they are hard to read right now.

Reply:
cleaned up!